### ISSUE TITLE
RFC: breaking changes in PouchDB 7.0.0

### ISSUE THREAD

**Issue by nolanlawson** [pouchdb/pouchdb#6019]
As Tom Dale [has noted](https://medium.com/@tomdale/semver-shouldnt-mean-more-breaking-changes-abfb7990dbab) it's easy to get into a habit of releasing breaking changes far too frequently and thus causing headaches for users. However it's been 3 months since 6.0.0, so it's a good time to at least start thinking about what we'd like to do in 7.0.0. (My preference is to pack as many breaking changes into a single major release, in order to reduce cognitive overhead for users trying to figure out how to upgrade from vX to vY, i.e. reduce the possible combinations of X and Y.)

@daleharvey and I have discussed this in https://github.com/pouchdb/pouchdb/pull/6003#issuecomment-266956515 and https://github.com/pouchdb/pouchdb/issues/5835 and elsewhere, but basically we'd like to focus on making 7.0.0 smaller and reducing excess baggage. Since we now have the plugin system this is fairly easy, however there are some caveats which I'd like to lay out in this RFC.

### pouchdb-core -> pouchdb-core, pouchdb-debug, and pouchdb-polyfills

So my basic idea is that we would add two new packages: `pouchdb-debug` and `pouchdb-polyfills`. `pouchdb-debug` would contain `debug` (2.5kb min+gz) and would rely on a generic event system from `pouchdb-core` in order to do its debugging. `pouchdb-polyfills` would contain built-in polyfills for Promise, fetch, Map, Set, Blob, Object.assign, and anything else we are currently polyfilling (I estimate this can save around ~5kb min+gz). This can be done in such a way that it doesn't muck with globals; e.g. by creating a `PouchDB._polyfills` object.

The challenge is that if we remove these things from `pouchdb-core` then it's technically a breaking change, because e.g. a project like [pouchdb-idb](https://github.com/gr2m/pouchdb-idb) would instantly break for anyone using `PouchDB.debug()` or relying on the polyfills. Hence 7.0.0.

An alternate idea is that we break up `pouchdb-core` into three new modules, `pouchdb-core-next`, `pouchdb-debug`, and `pouchdb-polyfills`, with the intention of eventually replacing `pouchdb-core` with `pouchdb-core-next`, but I worry that this might be too confusing and we might as well just do the 7.0.0 release sooner rather than later. But it is an option if we want to delay a major release.

Also to be clear: `pouchdb-polyfills` would still be a "default" plugin; I think it's way too premature to ship `dist/pouchdb.js` without the polyfills, because that build is designed for less-savvy users who don't want to have to worry about IE11 randomly breaking. `pouchdb-polyfills` is a "pro" feature for those who are comfortable with using browserify/webpack/etc. and using the custom builds. I strongly feel that we need a "just works" experience for newbies and a "build your own" experience for advanced users.

For `pouchdb-debug` I'm on the fence as to whether it should be a default plugin or not, but again it hardly matters since advanced users can always remove it.

### other stuff

Potentially we can also consider extracting other things like `PouchDB.defaults()` but it probably won't save a huge number of bytes.

I also can't think of anything else I'd want to deprecate for 7.0.0, but I'm sure @daleharvey and others have ideas.

**Comment by nolanlawson** [comment]
Somewhat related: https://github.com/pouchdb/pouchdb/issues/6020

**Comment by millette** [comment]
> Also to be clear: pouchdb-polyfills would still be a "default" plugin; I think it's way too premature to ship dist/pouchdb.js without the polyfills, because that build is designed for less-savvy users who don't want to have to worry about IE11 randomly breaking. pouchdb-polyfills is a "pro" feature for those who are comfortable with using browserify/webpack/etc. and using the custom builds. I strongly feel that we need a "just works" experience for newbies and a "build your own" experience for advanced users.

What about pouchdb for node? The polyfills probably aren't needed, right?

**Comment by nolanlawson** [comment]
Not necessarily! We support node 0.10 and above; features like Promises, Map, Set, Object.assign, etc aren't universally available.

**Comment by millette** [comment]
Although [node 0.10 and 0.12 are going out of commission by the end of the year](https://github.com/nodejs/LTS).

**Comment by nolanlawson** [comment]
Hm, just checked Node 4 though, and it wouldn't support our Map/Set usage (doesn't have 100% Map/Set support).

For Node though it's less of a drag to have polyfills included; this is mostly an issue for browser bundles which are more sensitive to size/deps

**Comment by broerse** [comment]
Perhaps we can make [point 6](https://pouchdb.com/2014/06/17/12-pro-tips-for-better-code-with-pouchdb.html) work out of the box with some checksum technique. Or put it under a feature flag.

**Comment by broerse** [comment]
Point 6 out of the box

> So if nothing changed in a document, don't bother put()ing it again.

would also help to implement this: https://github.com/nolanlawson/ember-pouch/issues/165

**Comment by nolanlawson** [comment]
@broerse Point 6 would add a lot of perf overhead and is arguably counterintuitive since it's not what CouchDB does. I'd prefer to have users do their own diffing. (It could also be done as a plugin, e.g. see pouchdb-upsert. :))

W.r.t. the `debug` and polyfills question, I have a new idea based on discussion in https://github.com/w3ctag/polyfills/issues/6:

- `PouchDB.debug()` is now a plugin.
- `require('pouchdb/no-polyfills')` provides a zero-polyfills version. We can accomplish this using Rollup and [rollup-plugin-inject](https://github.com/rollup/rollup-plugin-inject), and can also output a `dist/pouchdb.no-polyfills.js` UMD build. We can also promote [babel-preset-env](https://github.com/babel/babel-preset-env/) with the `useBuiltins` option as a way for users to automatically inject their own polyfills, using one set of polyfills for the entire codebase.

**Comment by nolanlawson** [comment]
TBH though I'm note sure exactly how `require('pouchdb/no-polyfills')` looks for all our other various packages (`pouchdb-browser`, `pouchdb-adapter-idb`, `pouchdb-utils`, etc.). Potentially each one needs its own `require('*/no-polyfills')`, but this can be done with build steps.

**Comment by broerse** [comment]
@nolanlawson Still think diffing should work out of the box because it creates less `_div`'s and less sync traffic.  So also more speed. That said a plugin would be the next best thing.

https://github.com/pouchdb/pouchdb/issues/6046 can also be considered breaking if you sync with an older PouchDB that doen not support `feed: 'normal'`

**Comment by nolanlawson** [comment]
Another thing I'd like to remove: `PouchDB.preferredAdapters` should be moved to `PouchDB.__preferredAdapters` to signal internal.

**Comment by nolanlawson** [comment]
Another nice-to-remove: `db.type()` (see https://github.com/pouchdb/pouchdb/issues/6106#issuecomment-277573228)

**Comment by nolanlawson** [comment]
We can also remove `pouchdb-mapreduce` as a default plugin. Once `pouchdb-find` is merged, we will have the necessary changes to the build script to ship standalone third-party plugins (i.e. with a `dist/` folder). The hardest part will just be updating documentation.

**Comment by nolanlawson** [comment]
I would be in favor of removing `PouchDB.defaults()`. It's incredibly buggy, we get bugs about it all the time (new one: https://github.com/nolanlawson/pouchdb-all-dbs/issues/25) and even folks like Hoodie that use it end up writing their own wrapper anyway due to all the bugs. It can exist as a separate plugin.

**Comment by nolanlawson** [comment]
`db.type()` can probably be removed in 7.0.0 if https://github.com/pouchdb/pouchdb/issues/6106 makes it in

**Comment by nolanlawson** [comment]
changes/replication filters would be nice to extract too: #6355

**Comment by nolanlawson** [comment]
I'm still not sure about which polyfills we should remove and when. We never really established at what point it makes sense to remove a polyfill – e.g. when 95% of users have it? 99%? etc.

I grabbed some quick data from [Paul Kinlan's tool](https://paulirish.github.io/web-feature-availability/) to give us an idea of what we can remove:

* Promise: 76%
* fetch: 61%
* Object.assign: can't find this data
* Map/Set: can't find this data

Since penetration of Object.assign and Map/Set are likely to be lower than that of Promise, I kind of feel like it's too early to remove any of these four. Also they wouldn't buy us much; I calculated that removing all of them would only save us 1.2KB: https://github.com/pouchdb/pouchdb/pull/6163#issuecomment-274241364
